include/dlaf/eigensolver/tridiag_solver/mc.h

```diff
@@ -122,28 +117,28 @@ void solveLeaf(Matrix<T, Device::CPU>& mat_trd, Matrix<T, Device::CPU>& mat_ev)
   }
 }
 
-template <class T>
-void copyDiagTile(const matrix::Tile<const T, Device::CPU>& tridiag_tile,
-                  const matrix::Tile<T, Device::CPU>& diag_tile) {
-  for (SizeType i = 0; i < tridiag_tile.size().rows(); ++i) {
-    diag_tile(TileElementIndex(i, 0)) = tridiag_tile(TileElementIndex(i, 0));
-  }
-}
-
-DLAF_MAKE_CALLABLE_OBJECT(copyDiagTile);
-DLAF_MAKE_SENDER_ALGORITHM_OVERLOADS(dlaf::internal::TransformDispatchType::Plain, copyDiagTile,
-                                     copyDiagTile_o)
-
-template <class T>
-void offloadDiagonal(Matrix<const T, Device::CPU>& mat_trd, Matrix<T, Device::CPU>& d) {
-  using dlaf::internal::Policy;
-  using dlaf::internal::whenAllLift;
-  using pika::execution::thread_priority;
-  using pika::execution::experimental::start_detached;
+template <class T, Device D>
+void offloadDiagonal(Matrix<const T, D>& mat_trd, Matrix<T, D>& evals) {
+  namespace ex = pika::execution::experimental;
+  namespace di = dlaf::internal;
 
-  for (SizeType i = 0; i < d.distribution().nrTiles().rows(); ++i) {
-    whenAllLift(mat_trd.read_sender(GlobalTileIndex(i, 0)), d.readwrite_sender(GlobalTileIndex(i, 0))) |
-        copyDiagTile(Policy<Backend::MC>(thread_priority::normal)) | start_detached();
+  for (SizeType i = 0; i < evals.distribution().nrTiles().rows(); ++i) {
+    auto diag_fn = [](const auto& tridiag_tile, const auto& diag_tile, [[maybe_unused]] auto&&... ts) {
+      if constexpr (D == Device::CPU) {
+        for (SizeType i = 0; i < tridiag_tile.size().rows(); ++i) {
+          diag_tile(TileElementIndex(i, 0)) = tridiag_tile(TileElementIndex(i, 0));
+        }
+      }
+      else {
+        copyDiagTileFromTridiagTile(tridiag_tile.size().rows(), tridiag_tile.ptr(), diag_tile.ptr(),
+                                    ts...);
+      }
```

Comment:
> this copies the diagnoal from one tile to another?
Yes.

> Would it be worth adding a copyDiagFromTridiagTile helper for the CPU case as well and then use the various macros to create all the needed overloads for this?

Yes, it would be worth adding, not just for this one but also for other custom kernels in `misg_gpu_kernels.h`. I plan to do that after the performance benchmarking. Either in this PR or in a new PR.

Reply:
I created an issue about this here : https://github.com/eth-cscs/DLA-Future/issues/634 . I am not sure if it would be better addressed in a separate PR.

Reply:
Separate PR is probably sanest, thanks!